dcap-ql/src/quote.rs

```diff
@@ -332,6 +452,84 @@ impl<'a> Qe3CertDataPpid<'a> {
     }
 }
 
+pub trait Quote3SignatureEcdsaP256Verifier {
+    /// Verify the platform certification data.
+    ///
+    /// The certification data is in `quote3signature.certification_data()`.
+    ///
+    /// On success, should return the platform certification public key (PCK).
+    // TODO: pass a container for the certification type and data instead of
+    // the whole signature structure.
+    fn verify_certification_data<'a>(&mut self, quote3signature: &'a Quote3SignatureEcdsaP256) -> Result<Cow<'a, [u8]>>;
+
+    /// Verify the quoting enclave (QE3).
+    fn verify_qe3(&mut self, qe3_report: &[u8], authentication_data: &[u8]) -> Result<()>;
+}
+
+pub trait Quote3SignatureVerify<'a>: Quote3Signature<'a> {
+    type TrustRoot;
+
+    fn verify(&self, quote: &[u8], root_of_trust: Self::TrustRoot) -> Result<()>;
+}
+
+#[cfg(feature = "verify")]
+impl<'a> Quote3SignatureVerify<'a> for Quote3SignatureEcdsaP256<'a> {
+    type TrustRoot = &'a mut dyn Quote3SignatureEcdsaP256Verifier;
+
+    fn verify(&self, quote: &[u8], verifier: Self::TrustRoot) -> Result<()> {
+        // Verify certification data
+        let pck = verifier.verify_certification_data(self)?;
+
+        // verify qe3_signature against qe3_report/authentication_data using pck
+        //   verify QE report signature signed by pck
+        let sig = self.qe3_signature();
+        let sig = get_ecdsa_sig_der(sig)?;
+        let data = self.qe3_report();
+        let mut hash = [0u8; 32];
+        Md::hash(hash::Type::Sha256, &data, &mut hash)?;
+        let pck = CString::new(pck)?;
+        let mut pck = Certificate::from_pem(pck.as_bytes_with_nul())?;
+        let pck_pk = pck.public_key_mut();
+        pck_pk.verify(mbedtls::hash::Type::Sha256, &hash, &sig)?;
+
+        //   verify QE report::reportdata
+        let mut qe3_report = Vec::with_capacity(Report::UNPADDED_SIZE);
+        qe3_report.extend(self.qe3_report());
+        qe3_report.resize_with(Report::UNPADDED_SIZE, Default::default);
+        let qe3_report = Report::try_copy_from(&qe3_report).ok_or(format_err!("Could not construct Qe3 report"))?;
+
+        let mut hash = [0u8; 32];
+        let att_key = self.raw_attestation_public_key();
```

Comment:
In a lot of places, you use a simple accessor on a struct and bind the result. Then you use the binding exactly once as a function parameter. Why?